pkgs/native_assets_builder/lib/src/utils/file.dart

```diff
@@ -50,4 +51,99 @@ extension DirectoryExtension on Directory {
     }
     return last;
   }
+
+  /// Run [callback] with this Dart process having exclusive access to this
+  /// [Directory].
+  ///
+  /// Note multiple isolates and isolate groups in the same Dart process share
+  /// locks, so these will be able to enter the exclusive section simultanously.
+  ///
+  /// The required [timeout] parameter determines how long to retry before
+  /// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+  // TODO(dacoharkes): Can we also make it in-process safe?
+  // TODO(dacoharkes): Where do we place this code so hook writers can reuse it?
+  // I don't really want to add it to package:native_assets_cli. A new package
+  // in dart-lang/tools?
+  Future<T> exclusive<T>(
```

Comment:
IMHO one should follow "separation of concerns" principle: This method chooses a file name and also performs the locking. I'd separate these two things.

Reply:
I've separated the implementation. But both the hook writers and `native_assets_builder` would duplicate code to write such locking file. So I've wrapped it in a function that still creates the lock file.